src/Storages/KeyDescriptionMonotonicityChecker.cpp

```diff
@@ -0,0 +1,249 @@
+                key_range.left_included = true;
+            }
+
+            if (!key_range.right.isNull())
+            {
+                key_range.right = applyFunction(func, current_type, key_range.right);
+                key_range.right_included = true;
+            }
+
+            current_type = func->getResultType();
+
+            if (!monotonicity.is_positive)
+                key_range.invert();
+        }
+
+        return monotonicity;
+    }
+
+    auto buildRPNContext(const KeyDescription & key_description, ContextPtr context)
+    {
+        auto partition_key_ast_clone = key_description.definition_ast->clone();
+
+        auto result = TreeRewriter(context).analyze(partition_key_ast_clone, key_description.expression->getRequiredColumnsWithTypes());
+
+        auto block_with_constants = KeyCondition::getBlockWithConstants(partition_key_ast_clone, result, context);
+
+        auto rpn_context = RPNBuilderTreeContext(context, block_with_constants, {});
+
+        return rpn_context;
+    }
+
+}
+
+IFunction::Monotonicity KeyDescriptionMonotonicityChecker::getMonotonicityInfo(
+    const KeyDescription & key_description,
+    const Range & range,
+    ContextPtr context
+)
+{
+    DataTypePtr key_column_type;
+
+    auto rpn_context = buildRPNContext(key_description, context);
+
+    auto rpn_function_list = buildRPNFunctionList(key_description, key_column_type, rpn_context);
+
+    auto function_list = buildFunctionList(rpn_function_list, key_column_type, context);
+
+    return applyMonotonicFunctionsChainToRange(range, function_list, key_column_type);
+}
+
+std::vector<RPNBuilderFunctionTreeNode> KeyDescriptionMonotonicityChecker::buildRPNFunctionList(const KeyDescription & key_description,
+                                                                                                DataTypePtr & key_expr_type,
+                                                                                                RPNBuilderTreeContext & rpn_context)
+{
+    auto rpn_node = RPNBuilderTreeNode(key_description.definition_ast.get(), rpn_context);
+
+    std::vector<RPNBuilderFunctionTreeNode> function_list;
+
+    buildRPNFunctionListImpl(rpn_node, key_description, key_expr_type, function_list);
+
+    return function_list;
+}
+
+std::vector<FunctionBasePtr> KeyDescriptionMonotonicityChecker::buildFunctionList(
+    const std::vector<RPNBuilderFunctionTreeNode> & rpn_function_list,
+    DataTypePtr key_column_type,
+    ContextPtr context
+)
+{
+    std::vector<FunctionBasePtr> functions;
+
+    for (auto it = rpn_function_list.rbegin(); it != rpn_function_list.rend(); ++it)
+    {
+        auto function = *it;
+        auto func_builder = FunctionFactory::instance().tryGet(function.getFunctionName(), context);
+
+        if (!func_builder)
+        {
+            throw Exception(ErrorCodes::BAD_ARGUMENTS, "Failed to check monotonicity of function {}", function.getFunctionName());
```

Comment:
Why could it fail? Should we use `get` instead of `tryGet` so it will throw `UNKNOWN_FUNCTION`?

Reply:
Possibly the same reason why it can fail here: https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/KeyCondition.cpp#L1481. I need to dive deep to remember / find out why it can fail.

`get` does make more sense